src/compressor/QatAccel.cc

```diff
@@ -11,67 +11,100 @@
  * Foundation.  See file COPYING.
  *
  */
-
+#include <qatzip.h>
 #include "QatAccel.h"
 
+#define SESSIONS_NUM 200
+
+void QzSessionDeleter::operator() (struct QzSession_S *session) {
+  if (NULL != session->internal) {
+    qzTeardownSession(session);
+    qzClose(session);
+  }
+  delete session;
+}
+
 /* Estimate data expansion after decompression */
 static const unsigned int expansion_ratio[] = {5, 20, 50, 100, 200};
 
-QatAccel::~QatAccel() {
-  if (NULL != session.internal) {
-    qzTeardownSession(&session);
-    qzClose(&session);
+QatAccel::QatAccel() {
+  sessions.resize(SESSIONS_NUM);
+  for (auto &session : sessions) {
+    session.reset(new struct QzSession_S);
+    memset(session.get(), 0, sizeof(struct QzSession_S));
   }
 }
 
+QatAccel::~QatAccel() {}
+
 bool QatAccel::init(const std::string &alg) {
   QzSessionParams_T params = {(QzHuffmanHdr_T)0,};
   int rc;
 
+  std::unique_lock<std::mutex> l(lock);
+  if (!alg.empty()) {
+    return true;
+  }
```

Comment:
```suggestion
  if (alg.empty()) {
    return true;
  }
```

and might want to move this check up before acquiring the lock?

Reply:
I originally wanted to prevent multiple calls during concurrency. But this logic seems wrong. I removed this judgment and used another way to prevent multiple calls during concurrency.